evm/db/chain.py

```diff
@@ -56,19 +57,56 @@ def get_canonical_head(self):
             raise CanonicalHeadNotFound("No canonical head set for this chain")
         return self.get_block_header_by_hash(self.db.get(CANONICAL_HEAD_HASH_DB_KEY))
 
-    def set_canonical_head(self, head_hash):
-        try:
-            self.get_block_header_by_hash(head_hash)
-        except BlockNotFound:
-            raise ValueError("Cannot use unknown block hash as canonical head: {}".format(
-                head_hash))
-        self.db.set(CANONICAL_HEAD_HASH_DB_KEY, head_hash)
+    def get_canonical_block_header_by_number(self, block_number):
+        """
+        Returns the block header with the given number in the canonical chain.
+
+        Raises BlockNotFound if there's no block header with the given number in the
+        canonical chain.
+        """
+        validate_uint256(block_number, title="Block Number")
+        return self.get_block_header_by_hash(self.lookup_block_hash(block_number))
 
     def get_score(self, block_hash):
         return rlp.decode(
             self.db.get(make_block_hash_to_score_lookup_key(block_hash)),
             sedes=rlp.sedes.big_endian_int)
 
+    def set_as_canonical_chain_head(self, header):
+        """
+        Sets the header as the canonical chain HEAD.
+        """
+        for h in reversed(self.find_common_ancestor(header)):
+            self.add_block_number_to_hash_lookup(h)
+
+        try:
+            self.get_block_header_by_hash(header.hash)
+        except BlockNotFound:
+            raise ValueError("Cannot use unknown block hash as canonical head: {}".format(
+                header.hash))
+        self.db.set(CANONICAL_HEAD_HASH_DB_KEY, header.hash)
+
+    @to_tuple
+    def find_common_ancestor(self, header):
+        """
+        Returns the chain leading up from the given header until the first ancestor it has in
+        common with our canonical chain.
+        """
+        h = header
+        while True:
+            yield h
+            if h.parent_hash == GENESIS_PARENT_HASH:
+                break
+            try:
+                orig = self.get_canonical_block_header_by_number(h.block_number)
+                if orig.hash == h.hash:
```

Comment:
Can this section get moved to an `else` clause to minimize what's present in the `try` block.

Reply:
Yep, sure